File: src/com/leozhi/topic116/Solution04.java
package com.leozhi.topic116;

/**
 * @author leozhi
 * 递归 + 连接两个节点
 * 通过
 * 2ms
 */
public class Solution04 {
    public Node connect(Node root) {
        if (root == null || root.left == null) {
            return root;
        }
        connectTwoNode(root.left, root.right);
        return root;
    }

    void connectTwoNode(Node node1, Node node2) {
        if (node1 == null) {
            return;
        }
        node1.next = node2;
        connectTwoNode(node1.left, node1.right);
        connectTwoNode(node1.right, node2.left);
        connectTwoNode(node2.left, node2.right);
    }
}

File: src/com/leozhi/topic206/Solution04.kt
package com.leozhi.topic206

/**
 * @author leozhi
 * 递归
 * 通过
 * 204ms
 */
class Solution04 {
    fun reverseList(head: ListNode?): ListNode? {
        if (head?.next == null) {
            return head
        }
        val reverseHead = reverseList(head.next)
        head.next.next = head
        head.next = null
        return reverseHead
    }
}